### ISSUE TITLE
Should `Hash` trait have ord/eq trait bounds

### ISSUE THREAD

**Issue by tcharding** [rust-bitcoin/rust-bitcoin#2500]
Are we happy with the `PartialOrd`, `Ord`, `PartialEq`, `Eq` trait bounds on the `Hash` trait. These could be moved to the `Bytes` associated type.

Conceptually the underlying byte array is the only thing that should be being compared but I'm not sure if making users call `as_byte_array()` everytime they want to compare is good ergonomics?

Originally mentioned: https://github.com/rust-bitcoin/rust-bitcoin/pull/2493#issuecomment-1959036085

**Comment by Kixunil** [comment]
> I'm not sure if making users call `as_byte_array()` everytime they want to compare is good ergonomics?

False dichotomy, they can require `Hash + Ord`.

`Eq` and `PartialEq` definitely sound good. The reason I doubted `PartialEq` is that some types may have different natural ordering for some reason (reversed types?)

**Comment by apoelstra** [comment]
> False dichotomy, they can require `Hash + Ord`.

Exactly -- having the bound on `Hash` is a pretty minor thing that basically saves the user typing (and maybe gatekeeps "what is a hash?" in some sense. But conversely if people implement the trait themselves, it requires extra boilerplate that maybe they don't need.

Personally I'm leaning a bit toward dropping the bound, on the basis that ordering hashes is a little obscure and it's better to give the user more control when implementing hashes and when using bounds on `Hash`.

For `PartialEq` and `Eq` I want to keep the bound. Comparing for equality is pretty-much the primary use case for hashes.

But I think we definitely want to implement `Ord` directly on the hashes so that people can put them into ordered collections, sort+dedup, etc.

> The reason I doubted `PartialEq` is that some types may have different natural ordering for some reason (reversed types?)

The reversing only ever applies to the hex display; and even if it were inherent to the data representation somehow, this wouldn't affect equality checking.

**Comment by Kixunil** [comment]
> But I think we definitely want to implement `Ord` directly on the hashes so that people can put them into ordered collections

We can drop the bounds and write `#[derive(Ord)]` in newtype macro calls in our crates.

> The reversing only ever applies to the hex display; and even if it were inherent to the data representation somehow, this wouldn't affect equality checking.

Sorry, I meant `PartialOrd`. I'm thinking more of a theoretical future hashes.

**Comment by tcharding** [comment]
> > But I think we definitely want to implement `Ord` directly on the hashes so that people can put them into ordered collections
> 
> We can drop the bounds and write `#[derive(Ord)]` in newtype macro calls in our crates.

We could also use `ordered::ArbitraryOrd`.

**Comment by 15IITian** [comment]
> We could also use ordered::ArbitraryOrd

Yeah I think , we should implement this instead of `Ord` trait as ordering the hashes of a particular instances does not sense logical but  have various use cases.
